internal/pkg/manifest/storage.go

```diff
@@ -0,0 +1,100 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package manifest
+
+import (
+	"errors"
+
+	"gopkg.in/yaml.v3"
+)
+
+var errUnmarshalEFSOpts = errors.New("unmarshal efs field into string or map")
+
+// StorageConfig is embedded into top level manifest structs when unmarshaling.
+type StorageConfig struct {
+	Storage Storage `yaml:"storage"`
+}
+
+// Storage represents the options for external and native storage.
+type Storage struct {
+	Volumes []Volume `yaml:"volumes"`
+}
+
+// Volume is an abstraction which merges the MountPoint and Volumes concepts from the ECS Task Definition
+type Volume struct {
+	Name           string        `yaml:"name"` // This is used to name the volume and fill the VolumeFrom field in the MountPoint
+	EFS            EFSIDOrConfig `yaml:"efs"`
+	MountPointOpts `yaml:",inline"`
+}
+
+// MountPointOpts is shared between Volumes for the main container and MountPoints for sidecars.
+type MountPointOpts struct {
+	ContainerPath string `yaml:"path"`
+	ReadOnly      bool   `yaml:"read_only"`
+}
+
+// MountPoint is used to let sidecars mount volumes defined in `storage`
+type MountPoint struct {
+	SourceVolume   string `yaml:"source_volume"`
+	MountPointOpts `yaml:",inline"`
+}
+
+// EFSIDOrConfig is a struct with a custom unmarshaler which can read either a string
+// or a more detailed struct with advanced options.
+type EFSIDOrConfig struct {
+	EFSID     *string
+	EFSConfig EFSVolumeConfiguration
```

Comment:
I've on second thought ripped out all the custom unmarshaling, because I believe EFS to be complex enough most customers will need more advanced config.